cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/tests/snowbridge.rs

```diff
@@ -535,3 +537,87 @@ fn send_token_from_ethereum_to_asset_hub_fail_for_insufficient_fund() {
 		assert_err!(send_inbound_message(make_register_token_message()), Token(FundsUnavailable));
 	});
 }
+
+// register weth as non-sufficient asset, dest account in this test does not exist so the transfer
+// should be trapped, just verify that the trapped origin contains the original sender so could
+// be claimed by himself from Gateway later.
+#[test]
+fn send_token_from_ethereum_to_asset_hub_verify_trapped_as_original_sender() {
+	BridgeHubRococo::fund_para_sovereign(AssetHubRococo::para_id().into(), INITIAL_FUND);
+
+	// Fund ethereum sovereign on AssetHub
+	AssetHubRococo::fund_accounts(vec![(AssetHubRococoReceiver::get(), INITIAL_FUND)]);
+
+	// Create asset on AssetHub.
+	let weth_asset_location: Location =
+		(Parent, Parent, EthereumNetwork::get(), AccountKey20 { network: None, key: WETH }).into();
+	let weth_asset_id: v3::Location = weth_asset_location.try_into().unwrap();
+	let asset_hub_sovereign = BridgeHubRococo::sovereign_account_id_of(Location::new(
+		1,
+		[Parachain(AssetHubRococo::para_id().into())],
+	));
+	AssetHubRococo::execute_with(|| {
+		assert_ok!(<AssetHubRococo as AssetHubRococoPallet>::ForeignAssets::force_create(
+			<AssetHubRococo as Chain>::RuntimeOrigin::root(),
+			weth_asset_id,
+			asset_hub_sovereign.into(),
+			false,
+			1,
+		));
+		assert!(<AssetHubRococo as AssetHubRococoPallet>::ForeignAssets::asset_exists(
+			weth_asset_id
+		));
+	});
+
+	BridgeHubRococo::execute_with(|| {
+		type RuntimeEvent = <BridgeHubRococo as Chain>::RuntimeEvent;
+		type EthereumInboundQueue =
+			<BridgeHubRococo as BridgeHubRococoPallet>::EthereumInboundQueue;
+		const XCM_FEE: u128 = 4_000_000_000;
+		let message_id: H256 = [0; 32].into();
+		let message = VersionedMessage::V1(MessageV1 {
+			chain_id: CHAIN_ID,
+			command: Command::SendToken {
+				token: WETH.into(),
+				destination: Destination::AccountId32 {
+					// dest account not exist and since weth is non-sufficient asset, the transfer
+					// should be trapped
+					id: [1; 32],
+				},
+				amount: 1_000_000_000,
+				fee: XCM_FEE,
+			},
+		});
+		let (xcm, _) = EthereumInboundQueue::do_convert(message_id, message).unwrap();
+		let _ = EthereumInboundQueue::send_xcm(xcm, AssetHubRococo::para_id().into()).unwrap();
+		// Check that the message was sent
+		assert_expected_events!(
+			BridgeHubRococo,
+			vec![
+				RuntimeEvent::XcmpQueue(cumulus_pallet_xcmp_queue::Event::XcmpMessageSent { .. }) => {},
+			]
+		);
+	});
+
+	AssetHubRococo::execute_with(|| {
+		type RuntimeEvent = <AssetHubRococo as Chain>::RuntimeEvent;
+
+		let expected_origin = Location {
+			parents: 2,
+			interior: Junctions::from([
+				GlobalConsensus(Ethereum { chain_id: 11155111 }),
+				AccountKey20 { network: None, key: [1u8; 20] },
+			]),
+		};
+
+		// Check that asset trapped as original sender
+		assert_expected_events!(
+			AssetHubRococo,
+			vec![
+				RuntimeEvent::PolkadotXcm(pallet_xcm::Event::AssetsTrapped { origin,.. }) => {
+					origin: *origin == expected_origin,
```

Comment:
Nice! Does this test pass?

And if so we should extent the test to actually claim an asset.

This also maybe useful in the future: https://github.com/paritytech/xcm-format/blob/master/proposals/0037-custom-asset-claimer.md

Reply:
Yeah, work as expected. Thanks for the sharing.

Reply:
https://github.com/Snowfork/polkadot-sdk/pull/121/commits/1ae2af6ad03381c6fa501f6bf22a7cc08b6c300d

The `send_token_from_ethereum_to_asset_hub_trapped_and_then_claimed` test works as expected including the   claim flow.